snuba/utils/snuba_exception.py

```diff
@@ -0,0 +1,99 @@
+from typing import Any, Dict, List, Optional, Type, TypedDict, Union, cast
+
+# mypy has not figured out recursive types yet so this can't be totally typesafe
+JsonSerializable = Union[str, int, float, bool, None, Dict[str, Any], List[Any]]
+
+
+class SnubaExceptionDict(TypedDict):
+    __type__: str
+    __name__: str
+    __message__: str
+    __extra_data__: Dict[str, JsonSerializable]
+    __should_report__: bool
+
+
+class _ExceptionRegistry:
+    """Keep a mapping of SnubaExceptions to their names"""
+
+    def __init__(self) -> None:
+        self.__mapping: Dict[str, Type["SnubaException"]] = {}
+
+    def register_class(self, cls: Type["SnubaException"]) -> None:
+        existing_class = self.__mapping.get(cls.__name__)
+        if not existing_class:
+            self.__mapping[cls.__name__] = cls
+
+    def get_class_by_name(self, cls_name: str) -> Optional[Type["SnubaException"]]:
+        return self.__mapping.get(cls_name)
+
+
+_REGISTRY = None
+
+
+def _get_registry() -> _ExceptionRegistry:
+    global _REGISTRY
+    # NOTE (Should this be protected by a mutex? I'm thinking it should be fine)
```

Comment:
I don't think this is thread safe. You would need a lock here to avoid multiple threads modifying the registry at the same time, since this will be called every time a `SnubaException` is thrown.

Reply:
I don't know if that's a performance problem since in theory exceptions happen very rarely. In a disaster scenario though this could make the disaster worse since not only are we throwing a lot of exceptions but those exceptions are bottlenecking on this class.

I think if you changed this a bit so the registry was created initially (instead of being set to None) and exception classes were added only once it would be better.

So if an exception happens, the registry is checked to see if that class has been added (which would be a threadsafe read). If the exception is not in the registry, then the lock is acquired and the class is added.

Reply:
This happens at the time the python AST is loaded into memory. The only time this would be modified by threads is if threads are defining a new exception locally inside themselves